spec/bolt/pal/yaml_plan_evaluator_spec.rb

```diff
@@ -0,0 +1,347 @@
+# frozen_string_literal: true
+
+require 'spec_helper'
+require 'bolt/pal'
+require 'bolt/pal/yaml_plan_evaluator'
+
+describe Bolt::PAL::YamlPlanEvaluator do
+  let(:plan_name) { Puppet::Pops::Loader::TypedName.new(:plan, 'test') }
+  let(:pal) { Bolt::PAL.new([], nil) }
+  # It doesn't really matter which loader or scope we use, but we need them, so take the
+  # static loader and global scope
+  let(:loader) { Puppet.lookup(:loaders).static_loader }
+  let(:scope) { Puppet.lookup(:global_scope) }
+
+  around :each do |example|
+    pal.in_bolt_compiler do
+      example.run
+    end
+  end
+
+  before :each do
+    # Make sure we don't accidentally call any run functions
+    allow(scope).to receive(:call_function)
+  end
+
+  def call_plan(plan, params = {})
+    # This is taken from the boltlib::run_plan function
+    catch(:return) {
+      plan.class.dispatcher.dispatchers[0].call_by_name_with_scope(scope, params, true)
+    }&.value
+  end
+
+  describe Bolt::PAL::YamlPlanEvaluator::PlanWrapper do
+    let(:plan) { described_class.new(plan_name, @plan_body) }
+
+    it 'parses parameter types' do
+      @plan_body = {
+        'parameters' => {
+          'nodes' => {
+            'type' => 'TargetSpec'
+          },
+          'package' => {
+            'type' => 'String'
+          },
+          'count' => {
+            'type' => 'Integer'
+          }
+        }
+      }
+
+      param_types = plan.parameters.inject({}) { |acc, param| acc.merge(param.name => param.type_expr) }
+
+      expect(param_types.keys).to contain_exactly('nodes', 'package', 'count')
+      expect(param_types['nodes'].name).to eq('TargetSpec')
+      expect(param_types['package']).to be_a(Puppet::Pops::Types::PStringType)
+      expect(param_types['count']).to be_a(Puppet::Pops::Types::PIntegerType)
+    end
+
+    it 'handles parameters without types' do
+      @plan_body = {
+        'parameters' => {
+          'empty' => {},
+          'nil' => nil
+        }
+      }
+
+      expect(plan.parameters.map(&:name)).to contain_exactly('empty', 'nil')
+      expect(plan.parameters.map(&:type_expr)).to eq([nil, nil])
+    end
+
+    it 'uses an empty parameter list if non are specified' do
+      @plan_body = {}
+
+      expect(plan.parameters).to eq([])
+    end
+  end
+
+  describe "::create" do
+    it 'fails if the plan is not a Hash' do
+      plan_body = '[]'
+
+      expect { described_class.create(loader, plan_name, 'test.yaml', plan_body) }.to raise_error(
+        ArgumentError, /test.yaml does not contain an object/
+      )
+    end
+
+    it 'returns a puppet function wrapper' do
+      plan_body = '{}'
+      plan = described_class.create(loader, plan_name, 'test.yaml', plan_body)
+      expect(plan).to be_a(Puppet::Functions::Function)
+    end
+  end
+
+  describe "evaluating a plan" do
```

Comment:
I'd like to see a test of variable extraction at this level in addition to the run step level.